### DISCUSSION TITLE
Flax submodule not showing up as trainable in Flax

### DISCUSSION THREAD

**Discussion by rashomon-gh** [google/flax#2245] [Q&A] [answered]
I have the following Flax module definition, where the huggingface BERT model is also a flax submodule:

```python
import flax
import jax
import jax.numpy as jnp
from flax import linen as nn
from transformers import AutoTokenizer, FlaxBertModel
from transformers.utils import logging


class Classifier(nn.Module):

    def setup(self):
        self.bert = FlaxBertModel.from_pretrained('bert-base-cased')
        self.fc = nn.Dense(features=2)

    def __call__(self, input_ids, attention_mask):
        out = self.bert(input_ids, attention_mask)

        out = out.pooler_output

        out = self.fc(out)
        out = jax.nn.log_softmax(out, axis=-1)

        return out
```

With an equivalent pytorch definition, bert params will be shared with the Classifier parameters  as trainable (unless marked as frozen in __init__). However in Flax, even after declaring bert in setup, I get the following set of params from the module pytree:

```
FrozenDict({
    params: {
        fc: {
            kernel: DeviceArray([[ 0.00646114,  0.00364647],
                         [ 0.07954237, -0.06435367],
                         [-0.01362615,  0.02644117],
                         ...,
                         [ 0.00158925,  0.02556718],
                         [-0.07334478,  0.03936611],
                         [-0.0254313 , -0.01591106]], dtype=float32),
            bias: DeviceArray([0., 0.], dtype=float32),
        },
    },
})
```

Am I missing something here or is there some Flax specific method to add bert to the pytree? For reference, I'm using the following to init the params.

```python
clf = Classifier()
key = jax.random.PRNGKey(0)

tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')
dummy_input = tokenizer.encode_plus(
        "This is some dummy text",
        add_special_tokens=True,
        max_length=512,
        truncation=True,
        return_token_type_ids=False,
        padding="max_length",
        return_attention_mask=True,
        return_tensors="np"
    )


params = clf.init(key, dummy_input["input_ids"], dummy_input["attention_mask"])
```

**Comment by jheek** [comment]
I'm not very familiar with the HF API but I think FlaxBertModel is not a subclass of linen.Module?
I guess FlaxBertModule is a wrapper around the actual Module and its parameters.

**Reply by rashomon-gh** [reply]
Well in that case, is there a way to share the parameters?

**Reply by rashomon-gh** [reply]
So according to their documentation it is a [Flax module](https://huggingface.co/docs/transformers/main/en/model_doc/bert#transformers.FlaxBertModel). So I'm a bit confused here. Is there a flax specific way to use one Flax module inside another ?

**Reply by marcvanzee** [reply]
You have to get the parameters out of the HF module and then insert them into Flax's variable dict. We actually discussed this with HuggingFace yesterday and it is a bit of a tricky case that we should try to simplify.

@cgarciae was interested in looking into this more.

**Reply by jheek** [reply]
> So according to their documentation it is a [Flax module]. So I'm a bit confused here

That's surprising indeed but I checked the source and it turns out it isn't really a module. There is a .module field that is a Flax Module.

I think it would be best to discuss this up with the HF team and ask them how you can incorporate their APIs into your own model

**Reply by rashomon-gh** [reply]
Thank you! I also found out yesterday that the HF model is a dict instead of being a FrozenDict. Pulling out all the params and then putting them inside a FrozenDict seems quite the work though!

**Reply by jheek** [reply]
You can feed dicts and FrozenDicts to flax. Calling FrozenDict(nested_dict) will freeze the whole nested dict at once so it's not a big issues to freeze something

**Reply by sanchit-gandhi** [reply]
Hey @ShawonAshraf!

If what you require is the `nn.Module` definition for Bert, you can access [`FlaxBertModule`](https://github.com/huggingface/transformers/blob/8f400775fc5bc1011a2674dcfd5408d30d69f678/src/transformers/models/bert/modeling_flax_bert.py#L924) by importing from `transformers.models.bert.modeling_flax_bert`.  The `FlaxBertModule` is a pure Flax `nn.Module`, so there should be no problems in using it for your example! You can then define `self.bert` as is done in other HF Flax Bert models (_c.f._ [`FlaxBertForPreTrainingModule`](https://github.com/huggingface/transformers/blob/8f400775fc5bc1011a2674dcfd5408d30d69f678/src/transformers/models/bert/modeling_flax_bert.py#L1008)):

```python
import flax
import jax
import jax.numpy as jnp
from flax import linen as nn
from transformers import AutoTokenizer, BertConfig
from transformers.models.bert.modeling_flax_bert import FlaxBertModule
from transformers.utils import logging

class Classifier(nn.Module):

    def setup(self):
        self.bert = FlaxBertModule(config=BertConfig())
        self.fc = nn.Dense(features=2)

    def __call__(self, input_ids, attention_mask):
        out = self.bert(input_ids, attention_mask)

        out = out.pooler_output

        out = self.fc(out)
        out = jax.nn.log_softmax(out, axis=-1)

        return out
```

If you require the pre-trained weights, putting the pre-trained params into the Flax variable dict is probably your best shout:
```python
from transformers import FlaxBertModel
from flax.core.frozen_dict import freeze, unfreeze

pretrained_bert = FlaxBertModel.from_pretrained('bert-base-cased')

params = unfreeze(params)

params['params']['bert'] = pretrained_bert.params

params = freeze(params)
```

**Reply by rashomon-gh** [reply]
Thanks @sanchit-gandhi , I'm actually looking to load the weights for fine-tuning.

**Reply by cgarciae** [reply]
Thanks @sanchit-gandhi! You beat me to it 😅
I'll write a full example below so people can copy-paste if needed.

**Comment by cgarciae** [comment, marked as answer]
For anyone interested I've cleaned up the code above, added some comments and made minor improvements, to create this full example on how to load a pretrained model from HuggingFace and correctly setup a Flax module for fine-tuning.

```python
import jax
from flax import linen as nn
from transformers import AutoTokenizer, FlaxBertModel
from flax.core.frozen_dict import freeze, unfreeze

class Classifier(nn.Module):
    bert: nn.Module

    def setup(self):
        self.fc = nn.Dense(features=2)

    def __call__(self, input_ids, attention_mask):
        out = self.bert(input_ids, attention_mask)
        out = out.pooler_output
        out = self.fc(out)
        return out

# load the pretrained huggingface model, this is not a Flax Module
pretrained_bert = FlaxBertModel.from_pretrained('bert-base-cased')
# create the Classifier module, extract the BERT Flax module from the huggingface model
clf = Classifier(pretrained_bert.module)

# create tokenizer
tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')
dummy_input = tokenizer.encode_plus(
        "This is some dummy text",
        add_special_tokens=True, max_length=512,
        truncation=True, return_token_type_ids=False,
        padding="max_length", return_attention_mask=True,
        return_tensors="np"
    )

# randomly initialize the module
key = jax.random.PRNGKey(0)
variables = clf.init(key, dummy_input["input_ids"], dummy_input["attention_mask"])

# replace randomly initialized params with pretrained params
variables = unfreeze(variables)
variables['params']['bert'] = pretrained_bert.params
variables = freeze(variables)

# run the module
preds = clf.apply(variables, dummy_input["input_ids"], dummy_input["attention_mask"])
print(preds)
```

It would be interesting if we expanded this into a guide on how to then setup optax to e.g. use different learning rates for between the `bert` backbone and the `fc` layer for fine-tuning.

**Reply by Yuji-github** [reply]
Why did you replace the randomly initialized params with the pre-trained params? 


variables = unfreeze(variables)
variables['params']['bert'] = pretrained_bert.params
variables = freeze(variables)

**Reply by rashomon-gh** [reply]
How else will you fine tune for a downstream task? Retraining the whole network certainly isn't efficient.

**Reply by cgarciae** [reply]
> How else will you fine tune for a downstream task? Retraining the whole network certainly isn't efficient. 

To freeze or set different learning rates to different subsets of weights you can use [optax.multi_transform](https://optax.readthedocs.io/en/latest/api.html#optax.multi_transform).

**Reply by Yuji-github** [reply]
@cgarciae and @ShawonAshraf Thank you for your advice first. 
@ShawonAshraf : Your are right, retraining the entire model is inefficient. 
@cgarciae : I will check it out the link and try it =)
